File: RAG/vector_store.py
"""Vector database for storing and retrieving multimodal embeddings."""

import os
import json
import pickle
from typing import List, Dict, Any, Optional, Tuple
import numpy as np
import chromadb
from chromadb.config import Settings
import config

class MultimodalVectorStore:
    """Vector database for storing text and image embeddings."""
    
    def __init__(self, persist_directory: str = None):
        """
        Initialize the vector store.
        
        Args:
            persist_directory: Directory to persist the database
        """
        self.persist_directory = persist_directory or config.VECTOR_DB_PATH
        os.makedirs(self.persist_directory, exist_ok=True)
        
        # Initialize ChromaDB
        self.client = chromadb.PersistentClient(path=self.persist_directory)
        
        # Create collections for different content types
        self.text_collection = self._get_or_create_collection("text_chunks")
        self.image_collection = self._get_or_create_collection("image_chunks")
        
        # Metadata storage for images (since ChromaDB doesn't store binary data well)
        self.image_metadata_file = os.path.join(self.persist_directory, "image_metadata.json")
        self.image_objects_file = os.path.join(self.persist_directory, "image_objects.pkl")
        
        self.image_metadata = self._load_image_metadata()
        self.image_objects = self._load_image_objects()
    
    def _get_or_create_collection(self, name: str):
        """Get or create a ChromaDB collection."""
        try:
            return self.client.get_collection(name)
        except ValueError:
            return self.client.create_collection(name)
    
    def _load_image_metadata(self) -> Dict[str, Any]:
        """Load image metadata from file."""
        if os.path.exists(self.image_metadata_file):
            with open(self.image_metadata_file, 'r') as f:
                return json.load(f)
        return {}
    
    def _save_image_metadata(self):
        """Save image metadata to file."""
        with open(self.image_metadata_file, 'w') as f:
            json.dump(self.image_metadata, f, indent=2)
    
    def _load_image_objects(self) -> Dict[str, Any]:
        """Load image objects from pickle file."""
        if os.path.exists(self.image_objects_file):
            with open(self.image_objects_file, 'rb') as f:
                return pickle.load(f)
        return {}
    
    def _save_image_objects(self):
        """Save image objects to pickle file."""
        with open(self.image_objects_file, 'wb') as f:
            pickle.dump(self.image_objects, f)
    
    def add_text_chunks(self, embedded_chunks: List[Dict[str, Any]]):
        """
        Add text chunks to the vector store.
        
        Args:
            embedded_chunks: List of embedded text chunks
        """
        if not embedded_chunks:
            return
        
        ids = [chunk['id'] for chunk in embedded_chunks]
        
        # Normalize embeddings if needed
        embeddings = []
        for chunk in embedded_chunks:
            emb = chunk['text_embedding']
            norm = np.linalg.norm(emb)
            if norm > 0:
                emb = emb / norm
            embeddings.append(emb.tolist())
            
        print(f"🔧 Adding {len(embeddings)} text embeddings (normalized)")
        
        # Prepare metadata (ChromaDB doesn't allow numpy arrays in metadata)
        metadatas = []
        documents = []
        
        for chunk in embedded_chunks:
            metadata = {
                'page_number': chunk['page_number'],
                'document_name': chunk['document_name'],
                'chunk_type': chunk['chunk_type'],
                'content_type': chunk['content_type']
            }
            # Add section_title if available
            if 'section_title' in chunk and chunk['section_title']:
                metadata['section_title'] = chunk['section_title'][:100]  # Truncate for metadata
            
            # Add all_sections if available (for enhanced section matching)
            if 'all_sections' in chunk and chunk['all_sections']:
                metadata['all_sections'] = ','.join(chunk['all_sections'])  # Store as comma-separated string
                #print(f"🔍 DEBUG: Storing all_sections in metadata: {metadata['all_sections']}")
            
            metadatas.append(metadata)
            documents.append(chunk['text'])
        
        # Add to ChromaDB
        self.text_collection.add(
            ids=ids,
            embeddings=embeddings,
            metadatas=metadatas,
            documents=documents
        )
        
        print(f"Added {len(embedded_chunks)} text chunks to vector store")
    
    def add_image_chunks(self, embedded_images: List[Dict[str, Any]]):
        """
        Add image chunks to the vector store.
        
        Args:
            embedded_images: List of embedded images
        """
        if not embedded_images:
            return
        
        ids = [img['id'] for img in embedded_images]
        # Always use image_embedding for consistency with existing collections
        embeddings = []
        for img in embedded_images:
            if 'image_embedding' in img:
                embeddings.append(img['image_embedding'].tolist())
            else:
                # This shouldn't happen, but fallback just in case
                raise ValueError(f"No image_embedding found for image {img.get('id', 'unknown')}")
        
        # Prepare metadata
        metadatas = []
        documents = []
        
        for img in embedded_images:
            metadata = {
                'page_number': img['page_number'],
                'document_name': img['document_name'],
                'caption': img['caption'],
                'ocr_text': img['ocr_text'],
                'content_type': img['content_type']
            }
            # Add section_title if available (always check, not just for rich context)
            if 'section_title' in img and img['section_title']:
                metadata['section_title'] = img['section_title'][:100]  # Truncate for metadata
            
            # Add rich context metadata if available
            if 'surrounding_text' in img:
                metadata.update({
                    'surrounding_text': img['surrounding_text'][:500],  # Truncate for metadata
                    'page_context': img['page_context'][:300] if 'page_context' in img else ''
                })
            
            metadatas.append(metadata)
            
            # Use rich context text if available, otherwise use combined_text
            document_text = img.get('rich_context_text', img.get('combined_text', ''))
            documents.append(document_text)
            
            # Store additional metadata and image object separately
            enhanced_metadata = {
                'page_number': img['page_number'],
                'document_name': img['document_name'],
                'caption': img['caption'],
                'ocr_text': img['ocr_text'],
                'combined_text': img.get('combined_text', ''),
                'text_embedding': img.get('text_embedding', img.get('simple_text_embedding', [])),
                'clip_text_embedding': img.get('clip_text_embedding', img.get('simple_clip_embedding', []))
            }
            # Add rich context data if available
            if 'rich_context_text' in img:
                enhanced_metadata.update({
                    'rich_context_text': img['rich_context_text'],
                    'rich_text_embedding': img['rich_text_embedding'].tolist() if 'rich_text_embedding' in img else [],
                    'rich_clip_embedding': img['rich_clip_embedding'].tolist() if 'rich_clip_embedding' in img else [],
                    'surrounding_text': img.get('surrounding_text', ''),
                    'section_title': img.get('section_title', ''),
                    'page_context': img.get('page_context', ''),
                    'preceding_text': img.get('preceding_text', ''),
                    'following_text': img.get('following_text', '')
                })
            
            # Convert embeddings to list for JSON serialization
            for key in ['text_embedding', 'clip_text_embedding', 'rich_text_embedding', 'rich_clip_embedding']:
                if key in enhanced_metadata and hasattr(enhanced_metadata[key], 'tolist'):
                    enhanced_metadata[key] = enhanced_metadata[key].tolist()
                    
            self.image_metadata[img['id']] = enhanced_metadata
            
            # Store PIL image object
            self.image_objects[img['id']] = img['image_object']
        
        # Add to ChromaDB
        self.image_collection.add(
            ids=ids,
            embeddings=embeddings,
            metadatas=metadatas,
            documents=documents
        )
        
        # Save metadata and objects
        self._save_image_metadata()
        self._save_image_objects()
        
        print(f"Added {len(embedded_images)} image chunks to vector store")
    
    def search_text(self, query_embedding: np.ndarray, top_k: int = 5, 
                   document_filter: Optional[str] = None) -> List[Dict[str, Any]]:
        """
        Search for similar text chunks.
        
        Args:
            query_embedding: Query embedding
            top_k: Number of results to return
            document_filter: Optional document name to filter by
            
        Returns:
            List of similar text chunks with metadata
        """
        where_clause = {}
        if document_filter:
            where_clause['document_name'] = document_filter
        
        results = self.text_collection.query(
            query_embeddings=[query_embedding.tolist()],
            n_results=top_k,
            where=where_clause if where_clause else None
        )
        
        # Format results
        formatted_results = []
        for i in range(len(results['ids'][0])):
            distance = results['distances'][0][i]
            similarity = max(0, 1 - distance)
            
            result = {
                'id': results['ids'][0][i],
                'text': results['documents'][0][i],
                'metadata': results['metadatas'][0][i],
                'distance': distance,
                'similarity_score': similarity,
                'content_type': 'text'
            }
            formatted_results.append(result)
        
        return formatted_results
    
    def search_images(self, query_embedding: np.ndarray, top_k: int = 5,
                     document_filter: Optional[str] = None,
                     page_filter: Optional[Dict] = None,
                     section_filter: Optional[str] = None) -> List[Dict[str, Any]]:
        """
        Search for similar images.
        
        Args:
            query_embedding: Query embedding (image embedding from CLIP)
            top_k: Number of results to return
            document_filter: Optional document name to filter by
            page_filter: Optional filter dict (e.g., {'page_number': 1})
            section_filter: Optional section title to filter by
            
        Returns:
            List of similar images with metadata
        """
        where_clause = {}
        
        # Build a list of conditions for the $and operator if multiple filters are present
        conditions = []
        
        if document_filter:
            conditions.append({"document_name": document_filter})
        
        if page_filter:
            conditions.append(page_filter)
        
        if section_filter:
            conditions.append({"section_title": section_filter})
        
        # Handle multiple filters with ChromaDB's format
        if len(conditions) > 1:
            where_clause = {"$and": conditions}
        elif len(conditions) == 1:
            where_clause = conditions[0]
        
        results = self.image_collection.query(
            query_embeddings=[query_embedding.tolist()],
            n_results=top_k,
            where=where_clause if where_clause else None
        )
        
        # Format results with additional metadata and image objects
        formatted_results = []
        for i in range(len(results['ids'][0])):
            image_id = results['ids'][0][i]
            
            result = {
                'id': image_id,
                'text': results['documents'][0][i],
                'metadata': results['metadatas'][0][i],
                'distance': results['distances'][0][i],
                'similarity_score': max(0, 1 - results['distances'][0][i]),  # Ensure non-negative
                'content_type': 'image'
            }
            
            # Add additional metadata if available
            if image_id in self.image_metadata:
                result.update(self.image_metadata[image_id])
            
            # Add image object if available
            if image_id in self.image_objects:
                result['image_object'] = self.image_objects[image_id]
            
            formatted_results.append(result)
        
        return formatted_results
    
    def search_images_by_section(self, query_embedding: np.ndarray, section_titles: List[str], 
                                top_k: int = 5, document_filter: Optional[str] = None) -> List[Dict[str, Any]]:
        """
        Search for images that belong to specific sections.
        
        Args:
            query_embedding: Query embedding (image embedding from CLIP)
            section_titles: List of section titles to filter by
            top_k: Number of results to return (maximum)
            document_filter: Optional document name to filter by
            
        Returns:
            List of images from the specified sections (may be less than top_k)
        """
        if not section_titles:
            return self.search_images(query_embedding, top_k, document_filter)
        
        # Search for images from each section and combine results
        all_section_results = []
        results_per_section = max(1, top_k // len(section_titles)) + 1  # Distribute top_k across sections
        
        for section_title in section_titles:
            # Use the new section_filter parameter for direct filtering
            section_results = self.search_images(
                query_embedding,
                results_per_section,
                document_filter,
                section_filter=section_title
            )
            
            # Add matched section info
            for result in section_results:
                result['matched_section'] = section_title
            
            all_section_results.extend(section_results)
        
        # Remove duplicates if any
        seen_ids = set()
        unique_results = []
        for result in all_section_results:
            result_id = result.get('id')
            if result_id and result_id not in seen_ids:
                seen_ids.add(result_id)
                unique_results.append(result)
        
        print(f"🔍 Section-based image search: Found {len(unique_results)} images from {len(section_titles)} sections")
        
        # Sort by similarity score and return up to top_k (may be less)
        unique_results.sort(key=lambda x: x.get('similarity_score', 0), reverse=True)
        return unique_results[:top_k]
    
    def search_multimodal(self, text_embedding: np.ndarray, image_embedding: np.ndarray,
                         top_k: int = 5, text_weight: float = 0.5,
                         document_filter: Optional[str] = None) -> List[Dict[str, Any]]:
        """
        Search for both text and images with weighted combination.
        
        Args:
            text_embedding: Text query embedding
            image_embedding: Image query embedding (from CLIP)
            top_k: Number of results to return
            text_weight: Weight for text results (0-1), image weight = 1 - text_weight
            document_filter: Optional document name to filter by
            
        Returns:
            Combined list of text and image results
        """
        # Get text and image results
        text_results = self.search_text(text_embedding, top_k * 2, document_filter)
        image_results = self.search_images(image_embedding, top_k * 2, document_filter)
        
        # Apply weights to similarity scores
        for result in text_results:
            result['weighted_score'] = result['similarity_score'] * text_weight
        
        for result in image_results:
            result['weighted_score'] = result['similarity_score'] * (1 - text_weight)
        
        # Combine and sort by weighted score
        combined_results = text_results + image_results
        combined_results.sort(key=lambda x: x['weighted_score'], reverse=True)
        
        return combined_results[:top_k]
    
    def get_document_list(self) -> List[str]:
        """Get list of all documents in the vector store."""
        # Get from text collection
        text_docs = set()
        try:
            text_results = self.text_collection.get()
            for metadata in text_results['metadatas']:
                text_docs.add(metadata['document_name'])
        except:
            pass
        
        # Get from image collection
        image_docs = set()
        try:
            image_results = self.image_collection.get()
            for metadata in image_results['metadatas']:
                image_docs.add(metadata['document_name'])
        except:
            pass
        
        return sorted(list(text_docs.union(image_docs)))
    
    def get_stats(self) -> Dict[str, Any]:
        """Get statistics about the vector store."""
        try:
            text_count = self.text_collection.count()
        except:
            text_count = 0
        
        try:
            image_count = self.image_collection.count()
        except:
            image_count = 0
        
        return {
            'total_text_chunks': text_count,
            'total_image_chunks': image_count,
            'total_documents': len(self.get_document_list()),
            'persist_directory': self.persist_directory
        }
    
    def delete_document(self, document_name: str):
        """
        Delete all chunks related to a specific document.
        
        Args:
            document_name: Name of the document to delete
        """
        # Delete from text collection
        try:
            text_results = self.text_collection.get(where={'document_name': document_name})
            if text_results['ids']:
                self.text_collection.delete(ids=text_results['ids'])
                print(f"Deleted {len(text_results['ids'])} text chunks for document: {document_name}")
        except Exception as e:
            print(f"Error deleting text chunks: {e}")
        
        # Delete from image collection
        try:
            image_results = self.image_collection.get(where={'document_name': document_name})
            if image_results['ids']:
                self.image_collection.delete(ids=image_results['ids'])
                
                # Also remove from metadata and objects
                for img_id in image_results['ids']:
                    if img_id in self.image_metadata:
                        del self.image_metadata[img_id]
                    if img_id in self.image_objects:
                        del self.image_objects[img_id]
                
                self._save_image_metadata()
                self._save_image_objects()
                
                print(f"Deleted {len(image_results['ids'])} image chunks for document: {document_name}")
        except Exception as e:
            print(f"Error deleting image chunks: {e}")
    
    def clear_all(self):
        """Clear all data from the vector store."""
        try:
            self.client.delete_collection("text_chunks")
            self.client.delete_collection("image_chunks")
            
            # Recreate collections
            self.text_collection = self._get_or_create_collection("text_chunks")
            self.image_collection = self._get_or_create_collection("image_chunks")
            
            # Clear metadata
            self.image_metadata = {}
            self.image_objects = {}
            self._save_image_metadata()
            self._save_image_objects()
            
            print("Cleared all data from vector store")
        except Exception as e:
            print(f"Error clearing vector store: {e}")

File: server/utils/EmbeddingEngines/multimodal/vector_store.js
const { ChromaClient } = require("chromadb");
const fs = require("fs");
const path = require("path");

/**
 * Multimodal Vector Store using ChromaDB
 *
 * Implements dual-collection storage:
 * - text_chunks: Text embeddings (768D/384D)
 * - image_chunks: Image embeddings (512D CLIP)
 *
 * Supports:
 * - Two-stage enhanced retrieval
 * - Section-priority image matching
 * - Context-aware scoring
 */
class MultimodalVectorStore {
  constructor(storagePath) {
    this.storagePath = storagePath;
    this.client = null;
    this.textCollection = null;
    this.imageCollection = null;
    this.initialized = false;
    this.initPromise = null;

    // External storage for image metadata and objects
    this.imageMetadataPath = path.join(storagePath, "image_metadata.json");
    this.imageObjectsPath = path.join(storagePath, "image_objects.json");

    this.imageMetadata = {};
    this.imageObjects = {};

    this.log("Initializing MultimodalVectorStore");
  }

  /**
   * Ensure store is initialized before operations
   */
  async #ensureInitialized() {
    if (this.initialized) return;
    if (this.initPromise) return this.initPromise;
    this.initPromise = this.#init();
    await this.initPromise;
  }

  log(text, ...args) {
    console.log(`\x1b[34m[MultimodalVectorStore]\x1b[0m ${text}`, ...args);
  }

  /**
   * Initialize ChromaDB client and collections
   */
  async #init() {
    try {
      // Ensure storage directory exists (for metadata files)
      if (!fs.existsSync(this.storagePath)) {
        fs.mkdirSync(this.storagePath, { recursive: true });
      }

      // Initialize ChromaDB client - use same endpoint as main ChromaDB provider
      // ChromaDB requires an HTTP endpoint, not a file path
      const chromaEndpoint = process.env.CHROMA_ENDPOINT || "http://localhost:8000";

      // Parse authentication header if provided
      const parseAuthHeader = (headerValue) => {
        const [type, value] = headerValue.split(" ");
        if (type === "Authorization") return { Authorization: value };
        return { [type]: value };
      };

      const clientConfig = {
        path: chromaEndpoint,
      };

      // Add authentication if configured
      if (process.env.CHROMA_API_HEADER && process.env.CHROMA_API_KEY) {
        clientConfig.fetchOptions = {
          headers: parseAuthHeader(
            `${process.env.CHROMA_API_HEADER} ${process.env.CHROMA_API_KEY}`
          ),
        };
      }

      this.client = new ChromaClient(clientConfig);

      this.log(`Connecting to ChromaDB at ${chromaEndpoint}`);

      // Get or create collections with unique names for multimodal system
      const textCollectionName = "multimodal_text_chunks";
      const imageCollectionName = "multimodal_image_chunks";

      // List existing collections to check if they exist
      const existingCollections = await this.client.listCollections();
      const existingNames = existingCollections.map(c => c.name);

      this.log(`Existing collections: [${existingNames.join(', ')}]`);

      // Get or create text collection - use getOrCreateCollection to avoid race conditions
      try {
        this.textCollection = await this.client.getOrCreateCollection({
          name: textCollectionName,
          metadata: { "hnsw:space": "cosine" },
        });
        this.log(`Ready: ${textCollectionName}`);
      } catch (error) {
        this.log(`Error with text collection: ${error.message}`);
        // Try to just get it if creation failed
        this.textCollection = await this.client.getCollection({
          name: textCollectionName,
        });
        this.log(`Retrieved existing: ${textCollectionName}`);
      }

      // Get or create image collection
      try {
        this.imageCollection = await this.client.getOrCreateCollection({
          name: imageCollectionName,
          metadata: { "hnsw:space": "cosine" },
        });
        this.log(`Ready: ${imageCollectionName}`);
      } catch (error) {
        this.log(`Error with image collection: ${error.message}`);
        // Try to just get it if creation failed
        this.imageCollection = await this.client.getCollection({
          name: imageCollectionName,
        });
        this.log(`Retrieved existing: ${imageCollectionName}`);
      }

      // Load external metadata
      this.#loadMetadata();

      this.initialized = true;
      this.log("ChromaDB initialized with dual collections");
    } catch (error) {
      this.log(`Initialization error: ${error.message}`);
      this.log(`Stack trace: ${error.stack}`);
      this.initialized = false;
      this.initPromise = null;
      throw error;
    }
  }

  /**
   * Load image metadata from file
   */
  #loadMetadata() {
    if (fs.existsSync(this.imageMetadataPath)) {
      this.imageMetadata = JSON.parse(
        fs.readFileSync(this.imageMetadataPath, "utf-8")
      );
    }

    if (fs.existsSync(this.imageObjectsPath)) {
      this.imageObjects = JSON.parse(
        fs.readFileSync(this.imageObjectsPath, "utf-8")
      );
    }
  }

  /**
   * Save image metadata to file
   */
  #saveMetadata() {
    fs.writeFileSync(
      this.imageMetadataPath,
      JSON.stringify(this.imageMetadata, null, 2)
    );

    fs.writeFileSync(
      this.imageObjectsPath,
      JSON.stringify(this.imageObjects, null, 2)
    );
  }

  /**
   * Add text chunks to vector store
   */
  async addTextChunks(textChunks, embeddings, documentName) {
    await this.#ensureInitialized();
    if (textChunks.length === 0) return;

    const ids = textChunks.map(chunk => chunk.id);
    const documents = textChunks.map(chunk => chunk.text);
    const metadatas = textChunks.map(chunk => ({
      documentName,
      pageNumber: chunk.pageNumber || 0,
      chunkIndex: chunk.chunkIndex,
      chunkType: "text",
      contentType: "text",
    }));

    await this.textCollection.add({
      ids,
      embeddings,
      documents,
      metadatas,
    });

    this.log(`Added ${textChunks.length} text chunks`);
  }

  /**
   * Add image chunks to vector store
   */
  async addImageChunks(images, embeddings, documentName) {
    await this.#ensureInitialized();
    if (images.length === 0) return;

    const ids = images.map(img => img.id);
    const documents = images.map(
      img => `${img.caption} ${img.ocrText} ${img.surroundingText}`.trim()
    );
    const metadatas = images.map(img => ({
      documentName,
      pageNumber: img.pageNumber || 0,
      caption: img.caption || "",
      ocrText: img.ocrText || "",
      contentType: "image",
    }));

    await this.imageCollection.add({
      ids,
      embeddings,
      documents,
      metadatas,
    });

    // Store additional metadata and image objects
    for (const img of images) {
      this.imageMetadata[img.id] = {
        documentName,
        pageNumber: img.pageNumber,
        caption: img.caption,
        ocrText: img.ocrText,
        surroundingText: img.surroundingText,
      };

      // Store image as base64 string
      if (img.image && Buffer.isBuffer(img.image)) {
        this.imageObjects[img.id] = img.image.toString("base64");
      }
    }

    this.#saveMetadata();
    this.log(`Added ${images.length} image chunks`);
  }

  /**
   * Search text chunks
   */
  async searchText(queryEmbedding, topK = 5, filter = {}) {
    await this.#ensureInitialized();
    const results = await this.textCollection.query({
      queryEmbeddings: [queryEmbedding],
      nResults: topK,
      where: Object.keys(filter).length > 0 ? filter : undefined,
    });

    return this.#formatResults(results, "text");
  }

  /**
   * Search image chunks
   */
  async searchImages(queryEmbedding, topK = 5, filter = {}) {
    await this.#ensureInitialized();
    const results = await this.imageCollection.query({
      queryEmbeddings: [queryEmbedding],
      nResults: topK,
      where: Object.keys(filter).length > 0 ? filter : undefined,
    });

    const formatted = this.#formatResults(results, "image");

    // Add extended metadata and image objects
    for (const result of formatted) {
      if (this.imageMetadata[result.id]) {
        Object.assign(result, this.imageMetadata[result.id]);
      }

      if (this.imageObjects[result.id]) {
        result.imageData = this.imageObjects[result.id];
      }
    }

    return formatted;
  }

  /**
   * Enhanced two-stage query
   * Stage 1: Text search for context
   * Stage 2: Context-aware image search
   */
  async enhancedQuery(queryText, textEmbedding, topK = 5) {
    await this.#ensureInitialized();
    this.log(`Enhanced query: "${queryText}"`);

    // Stage 1: Text retrieval for context
    const textResults = await this.searchText(textEmbedding, topK * 2);

    if (textResults.length === 0) {
      return {
        textResults: [],
        imageResults: [],
        totalResults: 0,
        mode: "enhanced",
      };
    }

    // Extract context from text results
    const contextPages = [...new Set(textResults.map(r => r.metadata.pageNumber))];
    const contextDocs = [...new Set(textResults.map(r => r.metadata.documentName))];

    this.log(`Context: ${contextPages.length} pages, ${contextDocs.length} documents`);

    // Stage 2: Context-aware image search
    // Search images from same pages
    const imageResults = [];
    for (const page of contextPages) {
      const pageImages = await this.searchImages(
        textEmbedding, // Use text embedding for cross-modal search
        Math.ceil(topK / contextPages.length),
        { pageNumber: page }
      );

      // Boost images from relevant pages
      pageImages.forEach(img => {
        img.similarityScore = (img.similarityScore || 0) + 0.2; // Page match bonus
        img.contextBoost = 0.2;
      });

      imageResults.push(...pageImages);
    }

    // Sort by boosted similarity
    imageResults.sort((a, b) => b.similarityScore - a.similarityScore);

    return {
      textResults: textResults.slice(0, topK),
      imageResults: imageResults.slice(0, topK),
      totalResults: textResults.length + imageResults.length,
      mode: "enhanced",
      context: {
        pages: contextPages,
        documents: contextDocs,
      },
    };
  }

  /**
   * Format ChromaDB results
   */
  #formatResults(results, contentType) {
    const formatted = [];

    if (!results.ids || results.ids.length === 0) return formatted;

    const ids = results.ids[0];
    const documents = results.documents[0];
    const metadatas = results.metadatas[0];
    const distances = results.distances[0];

    for (let i = 0; i < ids.length; i++) {
      formatted.push({
        id: ids[i],
        text: documents[i],
        metadata: metadatas[i],
        distance: distances[i],
        similarityScore: Math.max(0, 1 - distances[i]),
        contentType,
      });
    }

    return formatted;
  }

  /**
   * Delete document from both collections
   */
  async deleteDocument(documentName) {
    await this.#ensureInitialized();
    // Delete from text collection
    const textResults = await this.textCollection.get({
      where: { documentName },
    });
    if (textResults.ids.length > 0) {
      await this.textCollection.delete({
        ids: textResults.ids,
      });
    }

    // Delete from image collection
    const imageResults = await this.imageCollection.get({
      where: { documentName },
    });
    if (imageResults.ids.length > 0) {
      await this.imageCollection.delete({
        ids: imageResults.ids,
      });

      // Remove from external storage
      for (const id of imageResults.ids) {
        delete this.imageMetadata[id];
        delete this.imageObjects[id];
      }
      this.#saveMetadata();
    }

    this.log(`Deleted document: ${documentName}`);
  }

  /**
   * Get statistics
   */
  async getStats() {
    await this.#ensureInitialized();
    const textCount = await this.textCollection.count();
    const imageCount = await this.imageCollection.count();

    return {
      textChunks: textCount,
      imageChunks: imageCount,
      totalChunks: textCount + imageCount,
      storagePath: this.storagePath,
    };
  }
}

module.exports = {
  MultimodalVectorStore,
};
